xtm/xtm_api.cc

```diff
@@ -0,0 +1,282 @@
+	 */
+	int writefd;
+	/** Message queue, it's size must be power of two */
+	struct xtm_scsp_queue<union xtm_msg> queue;
+};
+
+/**
+ * Function gets num data pointers, previously saved in data field of
+ * union xtm_msg. Used in case send/recv xtm API pattern.
+ * @param[in] queue - xtm queue
+ * @param[out] data - array to save data pointers from xtm_msg union data field
+ * @param[in] num - count of elements to get
+ * @retval return the number of data pointers actually read.
+ */
+static inline unsigned
+xtm_queue_get_data(struct xtm_queue *xtm_queue,
+		   void **data, unsigned num)
+{
+	struct xtm_scsp_queue<union xtm_msg> *queue = &xtm_queue->queue;
+	unsigned i = 0;
+	unsigned new_read = queue->read;
+	unsigned queue_write =
+		__atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+
+	for (i = 0; i < num; i++) {
+		if (new_read == queue_write)
+			break;
+		data[i] = queue->buffer[new_read].data;
+		new_read = (new_read + 1) & (queue->len - 1);
+	}
+	__atomic_store_n(&queue->read, new_read, __ATOMIC_RELEASE);
+	return i;
+}
+
+/**
+ * Function executes all functions contained in all queue elements.
+ * Used in case dispatch/invoke xtm API pattern
+ * @param[in] queue - xtm queue
+ * @retval return count of executed functions
+ */
+static inline unsigned
+xtm_queue_execute_funs(struct xtm_queue *xtm_queue)
+{
+	struct xtm_scsp_queue<union xtm_msg> *queue = &xtm_queue->queue;
+	unsigned i = 0;
+	unsigned new_read = queue->read;
+	unsigned queue_write =
+		__atomic_load_n(&queue->write, __ATOMIC_ACQUIRE);
+
+	for(;;) {
+		if (new_read == queue_write)
+			break;
+		queue->buffer[new_read].fun(queue->buffer[new_read].fun_arg);
+		new_read = (new_read + 1) & (queue->len - 1);
+		i++;
+	}
+	__atomic_store_n(&queue->read, new_read, __ATOMIC_RELEASE);
+	return i;
+}
+
+static inline int
+xtm_queue_flush_pipe(struct xtm_queue *xtm_queue)
+{
+	unsigned char tmp[XTM_PIPE_SIZE];
+	ssize_t read_bytes;
+	while ((read_bytes = read(xtm_queue->readfd, tmp, sizeof(tmp))) < 0 &&
+	        errno == EINTR)
+		;
+	if (read_bytes <= 0 && errno != EAGAIN)
+		return -1;
+	return 0;
+}
+
+
+struct xtm_queue *
+xtm_create(unsigned size)
+{
+	int filedes[2];
+	struct xtm_queue *queue = (struct xtm_queue *)
+		malloc(sizeof(struct xtm_queue) +
+		       size * sizeof(union xtm_msg));
+	if (queue == NULL)
+		return NULL;
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	if ((filedes[0] = eventfd(0, 0)) < 0)
+#else
+	if (pipe(filedes) < 0)
+#endif
+		goto free_queue;
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	queue->readfd = queue->writefd = filedes[0];
+#else
+	queue->readfd = filedes[0];
+	queue->writefd = filedes[1];
+#endif
+
+	if (fcntl(queue->readfd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->writefd, F_SETFL, O_NONBLOCK) < 0)
+		goto close_fd;
+	if (xtm_scsp_queue_init(&queue->queue, size) < 0) {
+		errno = EINVAL;
+		goto close_fd;
+	}
+	return queue;
+
+close_fd:
+	close(queue->readfd);
+	if (queue->readfd != queue->writefd)
+		close(queue->writefd);
```

Comment:
The `close()` calls might override the `errno` value.

Reply:
Yes, i think function must save last errno value (You suggest to save errno which was before close and ignore errno of close function?)

Reply:
I mean that if, for example, fctnl() failed above, and then close() fails here, then the original errno is lost, and the user won't see the real error reason. Yes, I suggest to save errno before error handling, and restore it afterwards. So the user would see the first error, the original one.